src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -310,7 +315,24 @@ private AnalyzerConfigDocumentState GetRequiredAnalyzerConfigDocumentState(Docum
         => GetRequiredProjectState(documentId.ProjectId).AnalyzerConfigDocumentStates.GetRequiredState(documentId);
 
     public ProjectState? GetProjectState(ProjectId projectId)
-        => ProjectStates.TryGetValue(projectId, out var state) ? state : null;
+        => GetProjectState(ProjectStates, projectId);
+
+    /// <summary>
+    /// Searches for the project state with the specified project id in the given project states.
+    /// </summary>
+    /// <remarks>Requires the input array to be sorted by Id</remarks>
+    private static ProjectState? GetProjectState(ImmutableArray<ProjectState> sortedPojectStates, ProjectId projectId)
+    {
+        var index = sortedPojectStates.BinarySearch(projectId, static (projectState, projectId) => CompareProjectIds(projectState.Id, projectId));
```

Comment:
add a debug asser tthat sortedProjectStates is actually sorted?

Reply:
Debug is already slow enough, and this is called a bunch. I'll pass as it's asserted now during the Branch.